src/fsharp/vs/service.fs

```diff
@@ -1372,153 +1418,120 @@ module internal Parser =
                             errorCount <- errorCount + 1
-                      
+
                 match exn with
 #if EXTENSIONTYPING
-                | {Exception = (:? TypeProviderError as tpe)} ->
-                    tpe.Iter (fun e ->
-                        let newExn = {exn with Exception = e}
-                        report newExn
-                    )
+                | { Exception = (:? TypeProviderError as tpe) } -> tpe.Iter(fun e -> report { exn with Exception = e })
 #endif
                 | e -> report e
-      
-        let errorLogger = 
-            { new ErrorLogger("ErrorHandler") with 
+
+        let errorLogger =
+            { new ErrorLogger("ErrorHandler") with
                 member x.DiagnosticSink (exn, isError) = diagnosticSink (if isError then FSharpErrorSeverity.Error else FSharpErrorSeverity.Warning) exn
                 member x.ErrorCount = errorCount }
-      
-      
+
         // Public members
         member x.ErrorLogger = errorLogger
         member x.CollectedDiagnostics = errorsAndWarningsCollector.ToArray()
         member x.ErrorCount = errorCount
-        member x.TcConfig with set tc = tcConfig <- tc
+        member x.ErrorSeverityOptions with set opts = options <- opts
         member x.AnyErrors = errorCount > 0
 
+    let getLightSyntaxStatus fileName options =
+        let lower = String.lowercase fileName
+        let lightOnByDefault = List.exists (Filename.checkSuffix lower) FSharpLightSyntaxFileSuffixes
+        let lightSyntaxStatus = if lightOnByDefault then (options.LightSyntax <> Some false) else (options.LightSyntax = Some true)
+        LightSyntaxStatus(lightSyntaxStatus, true)
 
-    /// ParseOneFile builds all the information necessary to report errors, match braces and build scopes 
-    ///
-    /// projectSourceFiles is only used to compute isLastCompiland, and is ignored if Build.IsScript(mainInputFileName)  is true.
-    let ParseOneFile (ctok, source: string, matchBracesOnly: bool, reportErrors: bool, mainInputFileName: string, projectSourceFiles: string list, tcConfig: TcConfig) =
-
-          // This function requires the compilation thread because we install error handlers, whose callbacks must
-          // be invoked on the compilation thread, no other reason known to date.
-          // We should check whether those are "real" reasons - we could for example make collecting errors thread safe.
-          RequireCompilationThread ctok 
+    let createLexerFunction fileName options lexbuf (errHandler: ErrorHandler) =
+        let lightSyntaxStatus = getLightSyntaxStatus fileName options
 
-          // Initialize the error handler 
-          let errHandler = new ErrorHandler(reportErrors, mainInputFileName, tcConfig, source)
-
-          // Adding this new-line character at the end of the source seems odd but is required for some unit tests
-          let source = if source.Length = 0 || not (source.[source.Length - 1] = '\n') then source + "\n" else source 
```

Comment:
This is the line that needs to be re-instated, or at least dealt with in a separate PR.  What failures did you find this caused?

Thanks

Reply:
It was causing access to non-existent lines in R# documents, which led to runtime exceptions.
It is possible to make a workaround for this hack but it clearly looks like a bug when a parser parses not a source you passed it but some different string.

Reply:
I've reverted it and filed a separate issue.

Reply:
@auduchinok Yes, I agree. I've added the context I know to the linked bug.  On the bright side it could be worse - up until last year the same routine added **three** new line characters! :)